### PR TITLE
Reduce binary size and build time with split DWARF

### PR DESCRIPTION
**Pull request by foreseeable**
Commit Message: Reduce binary size and build time with split DWARF
Additional Description: When envoy build with compilation modes `opt` and `dbg`, the binary size is largely increased due to huge amounts of debug information. By splitting the debug information into two parts at compile time, we can reduce the total size of the object files processed by the linker. This can reduce the `envoy-static` target size from ~1211MB to ~392MB in compilation mode `opt`. Since those two modes are used in the CI scripts, this PR should also improve CI performance.
Risk Level: low
Testing: N/A
Docs Changes: N/A
Release Notes: Updated release notes for new changes to compilation mechanisms.

### CODE REVIEW

**Comment by foreseeable** [comment]
\cc @ahedberg

**Comment by mattklein123** [comment]
This is AWESOME. Thank you so much! I actually noticed that my machine has been churning a ton on the linking step.

**Comment by mattklein123** [comment]
Will defer to @lizan for the final seal of approval. :)

File: .bazelrc [hunk 1]
```diff
@@ -18,6 +18,8 @@ build --action_env=BAZEL_LINKOPTS=-lm
 build --host_javabase=@bazel_tools//tools/jdk:remote_jdk11
 build --javabase=@bazel_tools//tools/jdk:remote_jdk11
 build --enable_platform_specific_config
+build --fission=dbg,opt
+build --features=per_object_debug_info
```
Comment (lizan) [inline .bazelrc:22]: I think this is Linux only for now?

Comment (foreseeable) [reply .bazelrc]: this has been moved to build:linux now

File: bazel/README.md [hunk 2]
```diff
@@ -409,8 +409,11 @@ bazel build -c dbg //test/common/http:async_client_impl_test
 gdb bazel-bin/test/common/http/async_client_impl_test
 ```
 
-Without the `-c dbg` Bazel option at the end of the command line the test
-binaries will not include debugging symbols and GDB will not be very useful.
+We need to use `-c dbg` Bazel option to generate debugging symbols and without
+that GDB will not be very useful. The debugging symbols are stored as separate
+debugging information files (.dwo files). When using RBE to build, it's
+necessary to add `--remote_download_outputs=all` to download those `.dwo` file
```
Comment (lizan) [inline bazel/README.md:415]: Never recommend `--remote_download_outputs=all`, this is will be more build performance degradation than build without split dwarf, instead, just building `.dwp` target at same time would be enough.

Comment (foreseeable) [reply bazel/README.md:415]: I see.

Comment (foreseeable) [reply bazel/README.md]: I pushed the new version of README. Now it has included the commands to build `.dwp` file in the guide.

File: bazel/README.md [hunk 3]
```diff
@@ -406,11 +406,15 @@ be installed.
 
 ```
 bazel build -c dbg //test/common/http:async_client_impl_test
+bazel build //test/common/http:async_client_impl_test.dwp
```
Comment (lizan) [inline bazel/README.md:409]: This need to be `bazel build -c dbg //test/common/http:async_client_impl_test //test/common/http:async_client_impl_test.dwp`
building dwp without `-c dbg` will result empty dwp file.

Comment (foreseeable) [reply bazel/README.md]: Now `-c dbg` has been added

**Comment by foreseeable** [comment]
I also tested tools/stack_decode.py, it still works from the steps in `bazel/README.md`

**Comment by foreseeable** [comment]
@lizan I have added one line to release note. Can you have a look?

**Comment by yuzd** [comment]
@foreseeable  can you help with this issue?
I use devcontainer for debugger envoy,
the //source/exe:envoy-static.dwp file can not be used by gdb.

https://github.com/envoyproxy/envoy/issues/22788

**Pull request merged**